### PR TITLE
K8SPG-89 Add TLS secrets fields

### PR DESCRIPTION
**Pull request by nonemax**

### CODE REVIEW

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| demand-backup | failed |
| init-deploy | passed |
| data-migration-gcs | passed |
| scaling | passed |
| recreate | passed |
| affinity | passed |
| monitoring | passed |
| self-healing | passed |
| operator-self-healing | passed |
| clone-cluster | passed |
| scheduled-backup | failed |

commit: https://github.com/percona/percona-postgresql-operator/pull/123/commits/bc5fabc4c76926cc239ac87cca346793d3571a6d
image: `perconalab/percona-postgresql-operator:PR-123-bc5fabc4-pgo-apiserver`

image: `perconalab/percona-postgresql-operator:PR-123-bc5fabc4-pgo-event`

image: `perconalab/percona-postgresql-operator:PR-123-bc5fabc4-pgo-rmdata`

image: `perconalab/percona-postgresql-operator:PR-123-bc5fabc4-pgo-scheduler`

image: `perconalab/percona-postgresql-operator:PR-123-bc5fabc4-postgres-operator`

image: `perconalab/percona-postgresql-operator:PR-123-bc5fabc4-pgo-deployer`

**Comment by cap1984** [comment]
@nonemax please add tls check test

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| init-deploy | passed |
| scaling | passed |
| recreate | passed |
| affinity | passed |
| monitoring | passed |
| self-healing | passed |
| operator-self-healing | passed |
| clone-cluster | passed |
| demand-backup | failed |
| data-migration-gcs | passed |
| scheduled-backup | failed |

commit: https://github.com/percona/percona-postgresql-operator/pull/123/commits/bc5fabc4c76926cc239ac87cca346793d3571a6d
image: `perconalab/percona-postgresql-operator:PR-123-bc5fabc4-pgo-apiserver`

image: `perconalab/percona-postgresql-operator:PR-123-bc5fabc4-pgo-event`

image: `perconalab/percona-postgresql-operator:PR-123-bc5fabc4-pgo-rmdata`

image: `perconalab/percona-postgresql-operator:PR-123-bc5fabc4-pgo-scheduler`

image: `perconalab/percona-postgresql-operator:PR-123-bc5fabc4-postgres-operator`

image: `perconalab/percona-postgresql-operator:PR-123-bc5fabc4-pgo-deployer`

File: e2e-tests/tls-check/run [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+cluster="some-name"
+kubectl_bin apply \
+	-f $conf_dir/${cluster}-secrets.yml \
+	-f ${src_dir}/deploy/ssl-secrets.yaml \
+	-f $conf_dir/client.yml
+
+spinup_pgcluster "${cluster}" "${src_dir}/deploy/cr.yaml" "" "" "" "" "" "true"
+compare_psql_cmd "select-1" \
+	'SELECT datname,usename, ssl FROM pg_stat_ssl JOIN pg_stat_activity ON pg_stat_ssl.pid = pg_stat_activity.pid' \
+	"postgres:postgres_pass@${cluster}.${namespace}"
```
Comment (hors) [inline e2e-tests/tls-check/run:22]: I think we need to connect to postgres  via pgBouncer  and using certificate. And also we need to check not only creation of cluster with ssl but also its update  (enable/disable ssl on existing one ).

**Review by hors — CHANGES_REQUESTED** [review]

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| demand-backup | passed |
| init-deploy | passed |
| scaling | passed |
| recreate | passed |
| affinity | passed |
| monitoring | passed |
| self-healing | passed |
| scheduled-backup | failed |
| operator-self-healing | passed |
| clone-cluster | passed |

commit: https://github.com/percona/percona-postgresql-operator/pull/123/commits/76aba1647c2b4eb52d7456179087a4e1ea3ad081
image: `perconalab/percona-postgresql-operator:PR-123-76aba164-pgo-apiserver`

image: `perconalab/percona-postgresql-operator:PR-123-76aba164-pgo-event`

image: `perconalab/percona-postgresql-operator:PR-123-76aba164-pgo-rmdata`

image: `perconalab/percona-postgresql-operator:PR-123-76aba164-pgo-scheduler`

image: `perconalab/percona-postgresql-operator:PR-123-76aba164-postgres-operator`

image: `perconalab/percona-postgresql-operator:PR-123-76aba164-pgo-deployer`

File: e2e-tests/tls-check/run (hunk 1 above)
Comment (cap1984) [inline e2e-tests/tls-check/run:22]: I believe we need to test all four mods from https://www.postgresql.org/docs/8.4/libpq-connect.html#LIBPQ-CONNECT-SSLMODE

File: e2e-tests/tls-check/conf/client.yml [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+        - name: pg-client
+          image: perconalab/percona-distribution-postgresql:13.2
+          imagePullPolicy: Always
+          command:
+          - sleep
+          args:
+          - "100500"
+          volumeMounts:
+            - name: ca
+              mountPath: "/etc/tls"
```
Comment (cap1984) [inline e2e-tests/tls-check/conf/client.yml:25]: tmp/tls please

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| demand-backup | failed |
| init-deploy | passed |
| data-migration-gcs | passed |
| scaling | passed |
| recreate | passed |
| affinity | passed |
| monitoring | passed |
| self-healing | passed |
| operator-self-healing | passed |
| clone-cluster | passed |
| tls-check | passed |
| scheduled-backup | failed |

commit: https://github.com/percona/percona-postgresql-operator/pull/123/commits/6e409734c32869ca3df2cf426635a3cfc5423085
image: `perconalab/percona-postgresql-operator:PR-123-6e409734-pgo-apiserver`

image: `perconalab/percona-postgresql-operator:PR-123-6e409734-pgo-event`

image: `perconalab/percona-postgresql-operator:PR-123-6e409734-pgo-rmdata`

image: `perconalab/percona-postgresql-operator:PR-123-6e409734-pgo-scheduler`

image: `perconalab/percona-postgresql-operator:PR-123-6e409734-postgres-operator`

image: `perconalab/percona-postgresql-operator:PR-123-6e409734-pgo-deployer`

File: percona/controllers/pgcluster/pgcluster.go [hunk 3]
```diff
@@ -67,6 +69,51 @@ func Update(clientset kubeapi.Interface, newPerocnaPGCluster, oldPerocnaPGCluste
...
+	if err != nil {
+		return errors.Wrap(err, "change bouncer size to 0")
+	}
+
+	ctx := context.TODO()
+	for i := 0; i <= 30; i++ {
+		time.Sleep(5 * time.Second)
+		_, err := clientset.AppsV1().Deployments(newPerocnaPGCluster.Namespace).Get(ctx,
+			newPerocnaPGCluster.Name+"-pgbouncer", metav1.GetOptions{})
+		if err != nil && strings.Contains(err.Error(), "not found") {
```
Comment (egegunes) [inline percona/controllers/pgcluster/pgcluster.go:87]: Is it possible to check it with `kubeapi.IsNotFound`?

File: percona/controllers/pgcluster/pgcluster.go [hunk 4]
```diff
@@ -67,6 +69,51 @@ func Update(clientset kubeapi.Interface, newPerocnaPGCluster, oldPerocnaPGCluste
 	return nil
 }
 
+func UpdatePgBouncer(clientset *kubeapi.Client, newPerocnaPGCluster, oldPerocnaPGCluster *crv1.PerconaPGCluster) error {
```
Comment (egegunes) [inline percona/controllers/pgcluster/pgcluster.go:72]: typo `Perocna` on multiple locations in the function

File: percona/controllers/pgc/pgc.go [hunk 5]
```diff
@@ -265,16 +265,20 @@ func (c *Controller) onUpdate(oldObj, newObj interface{}) {
...
-		return
+	} else {
+		err = pgcluster.Update(c.Client, newCluster, oldCluster)
+		if err != nil {
+			log.Errorf("update perconapgcluster: update pgcluster: %s", err)
+			return
+		}
+		err = pgreplica.Update(c.Client, newCluster, oldCluster)
+		if err != nil {
+			log.Errorf("update perconapgcluster: update pgreplica: %s", err)
```
Comment (egegunes) [inline percona/controllers/pgc/pgc.go:276]: should we continue after logging the error?

**Review by egegunes — CHANGES_REQUESTED** [review]

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| demand-backup | passed |
| init-deploy | passed |
| data-migration-gcs | passed |
| scaling | passed |
| recreate | passed |
| affinity | passed |
| monitoring | passed |
| self-healing | passed |
| operator-self-healing | passed |
| scheduled-backup | passed |
| clone-cluster | passed |
| tls-check | passed |

commit: https://github.com/percona/percona-postgresql-operator/pull/123/commits/d2042da50e456f6c2060b42a799ce1f39c5bee5b
image: `perconalab/percona-postgresql-operator:PR-123-d2042da5-pgo-apiserver`

image: `perconalab/percona-postgresql-operator:PR-123-d2042da5-pgo-event`

image: `perconalab/percona-postgresql-operator:PR-123-d2042da5-pgo-rmdata`

image: `perconalab/percona-postgresql-operator:PR-123-d2042da5-pgo-scheduler`

image: `perconalab/percona-postgresql-operator:PR-123-d2042da5-postgres-operator`

image: `perconalab/percona-postgresql-operator:PR-123-d2042da5-pgo-deployer`

**Review by egegunes — APPROVED** [review]

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| demand-backup | passed |
| init-deploy | passed |
| data-migration-gcs | passed |
| scaling | passed |
| recreate | passed |
| affinity | passed |
| monitoring | passed |
| self-healing | passed |
| scheduled-backup | passed |
| operator-self-healing | passed |
| clone-cluster | passed |
| tls-check | passed |

commit: https://github.com/percona/percona-postgresql-operator/pull/123/commits/f4344f910a0bad5d35e7aab7a13b9eef2e8b3d6f
image: `perconalab/percona-postgresql-operator:PR-123-f4344f91-pgo-apiserver`

image: `perconalab/percona-postgresql-operator:PR-123-f4344f91-pgo-event`

image: `perconalab/percona-postgresql-operator:PR-123-f4344f91-pgo-rmdata`

image: `perconalab/percona-postgresql-operator:PR-123-f4344f91-pgo-scheduler`

image: `perconalab/percona-postgresql-operator:PR-123-f4344f91-postgres-operator`

image: `perconalab/percona-postgresql-operator:PR-123-f4344f91-pgo-deployer`

**Review by cap1984 — APPROVED** [review]

**Review by hors — APPROVED** [review]

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |

commit: https://github.com/percona/percona-postgresql-operator/pull/123/commits/c2322ddf811a1a9aff2abfd8e9b368ca2510aa1d
image: `perconalab/percona-postgresql-operator:PR-123-c2322ddf-pgo-apiserver`

image: `perconalab/percona-postgresql-operator:PR-123-c2322ddf-pgo-event`

image: `perconalab/percona-postgresql-operator:PR-123-c2322ddf-pgo-rmdata`

image: `perconalab/percona-postgresql-operator:PR-123-c2322ddf-pgo-scheduler`

image: `perconalab/percona-postgresql-operator:PR-123-c2322ddf-postgres-operator`

image: `perconalab/percona-postgresql-operator:PR-123-c2322ddf-pgo-deployer`

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| demand-backup | passed |
| init-deploy | failed |
| data-migration-gcs | passed |
| scaling | passed |
| recreate | failed |
| affinity | passed |
| monitoring | passed |
| self-healing | passed |
| operator-self-healing | passed |
| scheduled-backup | passed |
| clone-cluster | passed |
| tls-check | passed |

commit: https://github.com/percona/percona-postgresql-operator/pull/123/commits/c2322ddf811a1a9aff2abfd8e9b368ca2510aa1d
image: `perconalab/percona-postgresql-operator:PR-123-c2322ddf-pgo-apiserver`

image: `perconalab/percona-postgresql-operator:PR-123-c2322ddf-pgo-event`

image: `perconalab/percona-postgresql-operator:PR-123-c2322ddf-pgo-rmdata`

image: `perconalab/percona-postgresql-operator:PR-123-c2322ddf-pgo-scheduler`

image: `perconalab/percona-postgresql-operator:PR-123-c2322ddf-postgres-operator`

image: `perconalab/percona-postgresql-operator:PR-123-c2322ddf-pgo-deployer`

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| demand-backup | passed |
| scheduled-backup | passed |
| init-deploy | passed |
| scaling | passed |
| recreate | passed |
| affinity | passed |
| monitoring | passed |
| self-healing | passed |
| operator-self-healing | passed |
| clone-cluster | passed |
| tls-check | passed |
| data-migration-gcs | passed |

commit: https://github.com/percona/percona-postgresql-operator/pull/123/commits/a8dc9a2478344a1452ecd4a06fe93cb1650cd467
image: `perconalab/percona-postgresql-operator:PR-123-a8dc9a24-pgo-apiserver`

image: `perconalab/percona-postgresql-operator:PR-123-a8dc9a24-pgo-event`

image: `perconalab/percona-postgresql-operator:PR-123-a8dc9a24-pgo-rmdata`

image: `perconalab/percona-postgresql-operator:PR-123-a8dc9a24-pgo-scheduler`

image: `perconalab/percona-postgresql-operator:PR-123-a8dc9a24-postgres-operator`

image: `perconalab/percona-postgresql-operator:PR-123-a8dc9a24-pgo-deployer`

**Pull request merged**